File: app/lib/features/order/repositories/order_repository.dart
import 'package:dio/dio.dart';

import '../../../core/error/app_exception.dart';
import '../../../shared/models/api_response.dart';
import '../../../shared/models/page_response.dart';
import '../models/coupon_preview_request.dart';
import '../models/coupon_preview_response.dart';
import '../models/order_detail_response.dart';
import '../models/order_summary_response.dart';
import '../models/payment_url_response.dart';
import '../models/place_order_request.dart';

/// The typed client of the customer order/checkout/payment endpoints
/// (dto-spec §20). The coupon-preview call lives here because the coupon concern
/// is owned by the order module (architecture-spec §4) — there is no separate
/// `CouponRepository`; the VNPay payment-URL request lives here for the same
/// reason — payment is an order concern, not a separate `PaymentRepository`
/// (Sprint 13 Task 10).
///
/// It only calls the API, unwraps the [ApiResponse] envelope, and returns the
/// typed payload — throwing [AppException] on failure (the injected [Dio]'s
/// `ErrorInterceptor` has already mapped the transport error, and its
/// `AuthInterceptor` has attached the CUSTOMER bearer). It holds no business
/// logic and touches no storage or navigation, so every enveloped business error
/// (the `COUPON_*`, `PRODUCT_VARIANT_*`, `CART_ITEM_*`, `ADDRESS_*`, `ORDER_*`,
/// and `PAYMENT_*` codes of error-spec §8.6/§8.8/§8.9/§8.10/§8.11/§8.15)
/// propagates unchanged as an [AppException] carrying the registry code.
class OrderRepository {
  const OrderRepository(this._dio);

  static const String _couponValidatePath = '/api/v1/coupons/validate';
  static const String _ordersPath = '/api/v1/orders';

  final Dio _dio;

  /// `POST /coupons/validate` — the server-computed checkout summary for the
  /// selected cart items, optionally applying a coupon.
  Future<CouponPreviewResponse> previewCoupon(
    CouponPreviewRequest request,
  ) async {
    try {
      final response = await _dio.post<Map<String, dynamic>>(
        _couponValidatePath,
        data: request.toJson(),
      );
      final envelope = ApiResponse<CouponPreviewResponse>.fromJson(
        response.data!,
        (json) => CouponPreviewResponse.fromJson(json! as Map<String, dynamic>),
      );
      return envelope.data!;
    } on DioException catch (exception) {
      throw _asAppException(exception);
    }
  }

  /// `POST /orders` — place an order from the caller's selected cart items.
  Future<OrderDetailResponse> placeOrder(PlaceOrderRequest request) async {
    try {
      final response = await _dio.post<Map<String, dynamic>>(
        _ordersPath,
        data: request.toJson(),
      );
      return _unwrapOrder(response);
    } on DioException catch (exception) {
      throw _asAppException(exception);
    }
  }

  /// `GET /orders` — the caller's orders, paginated and most-recent-first as the
  /// server returns them.
  Future<PageResponse<OrderSummaryResponse>> getMyOrders({
    int page = 0,
    int size = 20,
  }) async {
    final query = <String, dynamic>{'page': page, 'size': size};
    try {
      final response = await _dio.get<Map<String, dynamic>>(
        _ordersPath,
        queryParameters: query,
      );
      final envelope = ApiResponse<PageResponse<OrderSummaryResponse>>.fromJson(
        response.data!,
        (json) => PageResponse<OrderSummaryResponse>.fromJson(
          json! as Map<String, dynamic>,
          (item) =>
              OrderSummaryResponse.fromJson(item! as Map<String, dynamic>),
        ),
      );
      return envelope.data!;
    } on DioException catch (exception) {
      throw _asAppException(exception);
    }
  }

  /// `GET /orders/{id}` — one of the caller's orders in full detail.
  Future<OrderDetailResponse> getMyOrder(int id) async {
    try {
      final response = await _dio.get<Map<String, dynamic>>('$_ordersPath/$id');
      return _unwrapOrder(response);
    } on DioException catch (exception) {
      throw _asAppException(exception);
    }
  }

  /// `POST /orders/{id}/cancel` — cancel one of the caller's `PENDING` orders,
  /// returning the cancelled order.
  Future<OrderDetailResponse> cancelOrder(int id) async {
    try {
      final response = await _dio.post<Map<String, dynamic>>(
        '$_ordersPath/$id/cancel',
      );
      return _unwrapOrder(response);
    } on DioException catch (exception) {
      throw _asAppException(exception);
    }
  }

  /// `POST /orders/{id}/payment` — request a signed VNPay sandbox payment URL
  /// for one of the caller's `PENDING`/`UNPAID`/`VNPAY` orders (Sprint 13
  /// Task 10).
  Future<PaymentUrlResponse> createPaymentUrl(int orderId) async {
    try {
      final response = await _dio.post<Map<String, dynamic>>(
        '$_ordersPath/$orderId/payment',
      );
      final envelope = ApiResponse<PaymentUrlResponse>.fromJson(
        response.data!,
        (json) => PaymentUrlResponse.fromJson(json! as Map<String, dynamic>),
      );
      return envelope.data!;
    } on DioException catch (exception) {
      throw _asAppException(exception);
    }
  }

  OrderDetailResponse _unwrapOrder(Response<Map<String, dynamic>> response) {
    final envelope = ApiResponse<OrderDetailResponse>.fromJson(
      response.data!,
      (json) => OrderDetailResponse.fromJson(json! as Map<String, dynamic>),
    );
    return envelope.data!;
  }

  AppException _asAppException(DioException exception) {
    final error = exception.error;
    if (error is AppException) {
      return error;
    }
    return const AppException(message: 'An unexpected error occurred');
  }
}

File: admin/src/features/order/repositories/order-repository.ts
import type { AxiosInstance } from 'axios';

import { AppError } from '@/core/error/app-error';
import type { ApiResponse } from '@/shared/types/api-response';
import type { PageResponse } from '@/shared/types/page-response';

import type { AdminOrderDetailResponse } from '../models/admin-order-detail-response';
import type { AdminOrderSummaryResponse } from '../models/admin-order-summary-response';
import type { OrderStatus } from '../models/order-status';
import type { UpdateOrderStatusRequest } from '../models/update-order-status-request';

const ADMIN_ORDERS_PATH = '/api/v1/admin/orders';
const ORDERS_PATH = '/api/v1/orders';
const UNEXPECTED_MESSAGE = 'An unexpected error occurred';

/** Optional server-driven paging, status filter, and order-code search for {@link OrderRepository.list}. */
export interface ListOrdersParams {
  readonly page?: number;
  readonly size?: number;
  readonly status?: OrderStatus;
  readonly orderCode?: string;
}

/**
 * The typed client of the two ADMIN order read endpoints plus the frozen
 * status-transition endpoint (dto-spec §20, sprint-12-plan Task 03) — the
 * React analog of the Flutter order repository (react-guidelines §Repository
 * Contract). Mirrors `ProductRepository` (Sprint 11) verbatim; there is no
 * create / update / delete method — orders are not admin-CRUD (sprint-12-plan
 * Design Decision 4).
 *
 * It only calls the API, unwraps the [ApiResponse] envelope, and returns the
 * typed payload — throwing [AppError] on failure (the single Axios instance's
 * error interceptor has already mapped the transport error). It holds no
 * business logic, no feature state, and touches no storage or navigation.
 */
export class OrderRepository {
  constructor(private readonly client: AxiosInstance) {}

  /**
   * `GET /admin/orders` — a page of orders for ADMIN management,
   * most-recent-first, optionally filtered by `status` and/or searched by an
   * `orderCode` fragment. Both are omitted from the request entirely when
   * unset — a blank `orderCode` is treated as unset, so the server never
   * receives a literal `orderCode=`.
   */
  async list(params?: ListOrdersParams): Promise<PageResponse<AdminOrderSummaryResponse>> {
    const orderCode =
      params?.orderCode !== undefined && params.orderCode.length > 0 ? params.orderCode : undefined;
    const response = await this.client.get<ApiResponse<PageResponse<AdminOrderSummaryResponse>>>(
      ADMIN_ORDERS_PATH,
      { params: { page: params?.page, size: params?.size, status: params?.status, orderCode } },
    );
    return this.unwrap(response.data.data);
  }

  /** `GET /admin/orders/{id}` — the full ADMIN detail of an order, regardless of owner. */
  async get(id: number): Promise<AdminOrderDetailResponse> {
    const response = await this.client.get<ApiResponse<AdminOrderDetailResponse>>(
      `${ADMIN_ORDERS_PATH}/${id}`,
    );
    return this.unwrap(response.data.data);
  }

  /**
   * `PATCH /orders/{id}/status` — advances an order's status. The frozen
   * ADMIN endpoint lives under the customer order path, not `/admin/orders`
   * — the read and write surfaces have different prefixes and this method
   * must not "tidy" that (sprint-12-plan Task 03).
   */
  async updateStatus(
    id: number,
    request: UpdateOrderStatusRequest,
  ): Promise<AdminOrderDetailResponse> {
    const response = await this.client.patch<ApiResponse<AdminOrderDetailResponse>>(
      `${ORDERS_PATH}/${id}/status`,
      request,
    );
    return this.unwrap(response.data.data);
  }

  private unwrap<T>(data: T | null | undefined): T {
    if (data === undefined || data === null) {
      throw new AppError({ message: UNEXPECTED_MESSAGE });
    }
    return data;
  }
}
